diff --git a/dist/quill.js b/dist/quill.js
index 811b3d080aa2c3b6bfee42a981c95bbe838872c0..1867e13f49de7c006b885f8eb85262bbe8d11963 100644
--- a/dist/quill.js
+++ b/dist/quill.js
@@ -4591,7 +4591,9 @@ var Keyboard = function (_Module) {
           format: _this2.quill.getFormat(range),
           offset: offset,
           prefix: prefixText,
-          suffix: suffixText
+          suffix: suffixText,
+          // https://github.com/quilljs/quill/commit/90269c7af9fc03e46be257490b2b0aec85d9ecc7
+          event: evt,
         };
         var prevented = bindings.some(function (binding) {
           if (binding.collapsed != null && binding.collapsed !== curContext.collapsed) return false;